File: backend/src/routes/systemyIt.js
'use strict';
const express = require('express');
const { body } = require('express-validator');
const db = require('../config/db');
const { authenticate } = require('../middleware/auth');
const { roleGuard } = require('../middleware/roleGuard');
const { validate } = require('../middleware/validate');
const { writeAudit } = require('../services/auditService');

const router = express.Router();
const IT = ['IT_ADMIN', 'SUPERADMIN'];

router.get('/', authenticate, async (req, res) => {
  const { tenant_id } = req.query;
  let q = `SELECT s.*, u.imie || ' ' || u.nazwisko as dodany_przez_nazwa,
                  COUNT(DISTINCT z.id) as liczba_zakresow
             FROM systemy_it s
             LEFT JOIN uzytkownicy u ON u.id = s.dodany_przez
             LEFT JOIN zakres_uprawnien z ON z.system_id = s.id
            WHERE s.aktywny = TRUE`;
  const params = [];
  if (tenant_id) { params.push(tenant_id); q += ` AND (s.tenant_id = $${params.length} OR s.tenant_id IS NULL)`; }
  q += ' GROUP BY s.id, u.imie, u.nazwisko ORDER BY s.nazwa';
  const { rows } = await db.query(q, params);
  res.json(rows);
});

router.get('/:id', authenticate, async (req, res) => {
  const { rows } = await db.query('SELECT * FROM systemy_it WHERE id=$1', [req.params.id]);
  if (!rows.length) return res.status(404).json({ error: 'System nie istnieje.' });

  const { rows: moduly } = await db.query('SELECT * FROM modul_systemu WHERE system_id=$1 ORDER BY nazwa', [req.params.id]);
  const { rows: zakresy } = await db.query(
    `SELECT z.*, m.nazwa as modul_nazwa FROM zakres_uprawnien z
       LEFT JOIN modul_systemu m ON m.id = z.modul_id
      WHERE z.system_id=$1 ORDER BY z.nazwa`,
    [req.params.id]
  );
  res.json({ ...rows[0], moduly, zakresy });
});

router.post('/', authenticate, roleGuard(...IT),
  [body('nazwa').notEmpty().trim(), body('poziom_krytycznosci').isIn(['NISKI','SREDNI','WYSOKI','KRYTYCZNY'])],
  validate,
  async (req, res) => {
    const { tenant_id, nazwa, opis, wlasciciel, poziom_krytycznosci } = req.body;
    const { rows } = await db.query(
      `INSERT INTO systemy_it (tenant_id, nazwa, opis, wlasciciel, poziom_krytycznosci, dodany_przez)
       VALUES ($1,$2,$3,$4,$5,$6) RETURNING *`,
      [tenant_id || null, nazwa, opis || null, wlasciciel || null, poziom_krytycznosci, req.user.userId]
    );
    await writeAudit({ userId: req.user.userId, username: req.user.username, rola: req.user.rola, akcja: 'SYSTEM_IT_CREATE', tabelaDocelowa: 'systemy_it', rekordId: rows[0].id, noweDane: rows[0] });
    res.status(201).json(rows[0]);
  }
);

router.put('/:id', authenticate, roleGuard(...IT), async (req, res) => {
  const { nazwa, opis, wlasciciel, poziom_krytycznosci, aktywny, tenant_id } = req.body;
  const { rows } = await db.query(
    `UPDATE systemy_it SET nazwa=$1, opis=$2, wlasciciel=$3, poziom_krytycznosci=$4,
       aktywny=$5, tenant_id=$6 WHERE id=$7 RETURNING *`,
    [nazwa, opis, wlasciciel, poziom_krytycznosci, aktywny !== false, tenant_id || null, req.params.id]
  );
  if (!rows.length) return res.status(404).json({ error: 'System nie istnieje.' });
  res.json(rows[0]);
});

// Modules CRUD
router.post('/:id/moduly', authenticate, roleGuard(...IT),
  [body('nazwa').notEmpty().trim()],
  validate,
  async (req, res) => {
    const { nazwa, opis } = req.body;
    const { rows } = await db.query(
      'INSERT INTO modul_systemu (system_id, nazwa, opis) VALUES ($1,$2,$3) RETURNING *',
      [req.params.id, nazwa, opis || null]
    );
    res.status(201).json(rows[0]);
  }
);

router.put('/moduly/:id', authenticate, roleGuard(...IT), async (req, res) => {
  const { nazwa, opis, aktywny } = req.body;
  const { rows } = await db.query(
    'UPDATE modul_systemu SET nazwa=$1, opis=$2, aktywny=$3 WHERE id=$4 RETURNING *',
    [nazwa, opis || null, aktywny !== false, req.params.id]
  );
  res.json(rows[0]);
});

// Scopes CRUD
router.post('/:id/zakresy', authenticate, roleGuard(...IT),
  [body('nazwa').notEmpty().trim()],
  validate,
  async (req, res) => {
    const { nazwa, opis, modul_id, uprzywilejowany } = req.body;
    const { rows } = await db.query(
      'INSERT INTO zakres_uprawnien (system_id, modul_id, nazwa, opis, uprzywilejowany) VALUES ($1,$2,$3,$4,$5) RETURNING *',
      [req.params.id, modul_id || null, nazwa, opis || null, uprzywilejowany || false]
    );
    res.status(201).json(rows[0]);
  }
);

router.put('/zakresy/:id', authenticate, roleGuard(...IT), async (req, res) => {
  const { nazwa, opis, modul_id, uprzywilejowany } = req.body;
  const { rows } = await db.query(
    'UPDATE zakres_uprawnien SET nazwa=$1, opis=$2, modul_id=$3, uprzywilejowany=$4 WHERE id=$5 RETURNING *',
    [nazwa, opis || null, modul_id || null, uprzywilejowany || false, req.params.id]
  );
  res.json(rows[0]);
});

module.exports = router;

File: backend-py/routes/systemy_it.py
from __future__ import annotations
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel
from sqlalchemy import text
from sqlalchemy.orm import Session

from database import get_session
from dependencies import CurrentUser, get_current_user, require_roles, write_audit, get_client_ip

router = APIRouter()
IT = ("IT_ADMIN", "SUPERADMIN")


class SystemIn(BaseModel):
    tenant_id: Optional[int] = None
    nazwa: str
    opis: Optional[str] = None
    wlasciciel: Optional[str] = None
    poziom_krytycznosci: str = "NORMALNY"
    aktywny: Optional[bool] = True


class ModulIn(BaseModel):
    nazwa: str
    opis: Optional[str] = None
    aktywny: Optional[bool] = True


class ZakresIn(BaseModel):
    nazwa: str
    opis: Optional[str] = None
    modul_id: Optional[int] = None
    uprzywilejowany: Optional[bool] = False


@router.get("/")
async def list_systemy(tenant_id: Optional[int] = None,
                        session: Session = Depends(get_session),
                        current_user: CurrentUser = Depends(get_current_user)):
    q = """SELECT s.*, u.imie || ' ' || u.nazwisko as dodany_przez_nazwa,
                  COUNT(DISTINCT z.id) as liczba_zakresow
             FROM systemy_it s
             LEFT JOIN uzytkownicy u ON u.id=s.dodany_przez
             LEFT JOIN zakres_uprawnien z ON z.system_id=s.id
            WHERE s.aktywny=1"""
    params = {}
    if tenant_id:
        q += " AND (s.tenant_id=:tid OR s.tenant_id IS NULL)"
        params["tid"] = tenant_id
    q += " GROUP BY s.id, u.imie, u.nazwisko ORDER BY s.nazwa"
    rows = session.execute(text(q), params).mappings().all()
    return [dict(r) for r in rows]


@router.get("/{system_id}")
async def get_system(system_id: int, session: Session = Depends(get_session),
                      current_user: CurrentUser = Depends(get_current_user)):
    row = session.execute(text("SELECT * FROM systemy_it WHERE id=:id"), {"id": system_id}).mappings().first()
    if not row:
        raise HTTPException(status_code=404, detail="System nie istnieje.")
    result = dict(row)
    moduly = session.execute(text("SELECT * FROM modul_systemu WHERE system_id=:id ORDER BY nazwa"), {"id": system_id}).mappings().all()
    zakresy = session.execute(
        text("SELECT z.*, m.nazwa as modul_nazwa FROM zakres_uprawnien z LEFT JOIN modul_systemu m ON m.id=z.modul_id WHERE z.system_id=:id ORDER BY z.nazwa"),
        {"id": system_id},
    ).mappings().all()
    result["moduly"] = [dict(m) for m in moduly]
    result["zakresy"] = [dict(z) for z in zakresy]
    return result


@router.post("/", status_code=201)
async def create_system(body: SystemIn, request: Request,
                         session: Session = Depends(get_session),
                         current_user: CurrentUser = Depends(require_roles(*IT))):
    from datetime import datetime
    row = session.execute(
        text("INSERT INTO systemy_it (tenant_id,nazwa,opis,wlasciciel,poziom_krytycznosci,dodany_przez,data_dodania) VALUES (:tid,:n,:o,:w,:pk,:dp,:dd) RETURNING *"),
        {"tid": body.tenant_id, "n": body.nazwa, "o": body.opis, "w": body.wlasciciel,
         "pk": body.poziom_krytycznosci, "dp": current_user.userId, "dd": datetime.utcnow()},
    ).mappings().first()
    session.commit()
    result = dict(row)
    write_audit(session, user_id=current_user.userId, username=current_user.username,
                rola=current_user.rola, akcja="SYSTEM_IT_CREATE", tabela_docelowa="systemy_it",
                rekord_id=result["id"], nowe_dane=result, ip_adres=get_client_ip(request))
    return result


@router.put("/{system_id}")
async def update_system(system_id: int, body: SystemIn, session: Session = Depends(get_session),
                         current_user: CurrentUser = Depends(require_roles(*IT))):
    row = session.execute(
        text("UPDATE systemy_it SET nazwa=:n, opis=:o, wlasciciel=:w, poziom_krytycznosci=:pk, aktywny=:a, tenant_id=:tid WHERE id=:id RETURNING *"),
        {"n": body.nazwa, "o": body.opis, "w": body.wlasciciel, "pk": body.poziom_krytycznosci,
         "a": body.aktywny if body.aktywny is not None else True, "tid": body.tenant_id, "id": system_id},
    ).mappings().first()
    session.commit()
    if not row:
        raise HTTPException(status_code=404, detail="System nie istnieje.")
    return dict(row)


@router.post("/{system_id}/moduly", status_code=201)
async def create_modul(system_id: int, body: ModulIn, session: Session = Depends(get_session),
                        current_user: CurrentUser = Depends(require_roles(*IT))):
    row = session.execute(
        text("INSERT INTO modul_systemu (system_id,nazwa,opis) VALUES (:sid,:n,:o) RETURNING *"),
        {"sid": system_id, "n": body.nazwa, "o": body.opis},
    ).mappings().first()
    session.commit()
    return dict(row)


@router.put("/moduly/{modul_id}")
async def update_modul(modul_id: int, body: ModulIn, session: Session = Depends(get_session),
                        current_user: CurrentUser = Depends(require_roles(*IT))):
    row = session.execute(
        text("UPDATE modul_systemu SET nazwa=:n, opis=:o, aktywny=:a WHERE id=:id RETURNING *"),
        {"n": body.nazwa, "o": body.opis, "a": body.aktywny if body.aktywny is not None else True, "id": modul_id},
    ).mappings().first()
    session.commit()
    return dict(row)


@router.post("/{system_id}/zakresy", status_code=201)
async def create_zakres(system_id: int, body: ZakresIn, session: Session = Depends(get_session),
                         current_user: CurrentUser = Depends(require_roles(*IT))):
    row = session.execute(
        text("INSERT INTO zakres_uprawnien (system_id,modul_id,nazwa,opis,uprzywilejowany) VALUES (:sid,:mid,:n,:o,:u) RETURNING *"),
        {"sid": system_id, "mid": body.modul_id, "n": body.nazwa, "o": body.opis, "u": body.uprzywilejowany or False},
    ).mappings().first()
    session.commit()
    return dict(row)


@router.put("/zakresy/{zakres_id}")
async def update_zakres(zakres_id: int, body: ZakresIn, session: Session = Depends(get_session),
                         current_user: CurrentUser = Depends(require_roles(*IT))):
    row = session.execute(
        text("UPDATE zakres_uprawnien SET nazwa=:n, opis=:o, modul_id=:mid, uprzywilejowany=:u WHERE id=:id RETURNING *"),
        {"n": body.nazwa, "o": body.opis, "mid": body.modul_id, "u": body.uprzywilejowany or False, "id": zakres_id},
    ).mappings().first()
    session.commit()
    return dict(row)
